File: leetcode/1493. 删掉一个元素以后全为 1 的最长子数组.go
func longestSubarray(nums []int) int {
	n := len(nums)
	pre := make([]int, n)
	suf := make([]int, n)
	pre[0] = nums[0]
	for i := 1; i < n; i++ {
		if nums[i] != 0 {
			pre[i] = pre[i-1] + 1
		} else {
			pre[i] = 0
		}
	}
	suf[n-1] = nums[n-1]
	for i := n - 2; i >= 0; i-- {
		if nums[i] != 0 {
			suf[i] = suf[i+1] + 1
		} else {
			suf[i] = 0
		}
	}

	ans := 0
	for i := 0; i < n; i++ {
		preSum := 0
		if i != 0 {
			preSum = pre[i-1]
		}
		sufSum := 0
		if i != n-1 {
			sufSum = suf[i+1]
		}
		if preSum+sufSum > ans {
			ans = preSum + sufSum
		}
	}

	return ans
}

File: leetcode/1493. 删掉一个元素以后全为 1 的最长子数组.py
class Solution:
    def longestSubarray(self, nums: List[int]) -> int:
        n = len(nums)
        pre = [0] * n
        suf = [0] * n
        pre[0] = nums[0]
        for i in range(1, n):
            pre[i] = pre[i - 1] + 1 if nums[i] else 0
        suf[-1] = nums[-1]
        for i in range(n - 2, -1, -1):
            suf[i] = suf[i + 1] + 1 if nums[i] else 0

        ans = 0
        for i in range(n):
            pre_sum = pre[i - 1] if i > 0 else 0
            suf_sum = suf[i + 1] if i < n - 1 else 0
            ans = max(ans, pre_sum + suf_sum)

        return ans